src/Xamarin.Android.Build.Tasks/Tasks/CompileNativeAssembly.cs

```
@@ -0,0 +1,211 @@
+
+			string assemblerName = Path.GetFileName (config.AssemblerPath);
+			LogDebugMessage ($"[Native Assembler] {psi.FileName} {psi.Arguments}");
+			using (var proc = new Process ()) {
+				proc.OutputDataReceived += (s, e) => {
+					if (e.Data != null)
+						OnOutputData (assemblerName, s, e);
+					else
+						stdout_completed.Set ();
+				};
+
+				proc.ErrorDataReceived += (s, e) => {
+					if (e.Data != null)
+						OnErrorData (assemblerName, s, e);
+					else
+						stderr_completed.Set ();
+				};
+
+				proc.StartInfo = psi;
+				proc.Start ();
+				proc.BeginOutputReadLine ();
+				proc.BeginErrorReadLine ();
+				Token.Register (() => { try { proc.Kill (); } catch (Exception) { } });
+				proc.WaitForExit ();
+
+				if (psi.RedirectStandardError)
+					stderr_completed.WaitOne (TimeSpan.FromSeconds (30));
+
+				if (psi.RedirectStandardOutput)
+					stdout_completed.WaitOne (TimeSpan.FromSeconds (30));
+
+				return proc.ExitCode == 0;
+			}
+		}
+
+		IEnumerable<Config> GetAssemblerConfigs ()
+		{
+			string sdkBinDirectory = MonoAndroidHelper.GetOSBinPath ();
+			foreach (ITaskItem item in Sources) {
+				string abi = item.GetMetadata ("abi");
+				string prefix = String.Empty;
+				AndroidTargetArch arch;
+
+				switch (abi) {
+					case "armeabi-v7a":
+						prefix = Path.Combine (sdkBinDirectory, "arm-linux-androideabi");
+						arch = AndroidTargetArch.Arm;
+						break;
+
+					case "arm64":
+					case "arm64-v8a":
+					case "aarch64":
+						prefix = Path.Combine (sdkBinDirectory, "aarch64-linux-android");
+						arch = AndroidTargetArch.Arm64;
+						break;
+
+					case "x86":
+						prefix = Path.Combine (sdkBinDirectory, "i686-linux-android");
+						arch = AndroidTargetArch.X86;
+						break;
+
+					case "x86_64":
+						prefix = Path.Combine (sdkBinDirectory, "x86_64-linux-android");
+						arch = AndroidTargetArch.X86_64;
+						break;
+
+					default:
+						throw new Exception ("Unsupported Android target architecture ABI: " + abi);
```

We have string interpolation.  We should prefer it.

Also, we should throw a better exception type than `Exception`.

```suggestion
						throw new NotSupportedException ($"Unsupported Android target architecture ABI `{abi}`!");
```

---

copy pasta